### ISSUE TITLE
Possible to single source file globs and config options for reuse?

### ISSUE THREAD

**Issue by kevinSuttle** [prettier/prettier#4420]
Is there a way to define and reuse filepaths? 

## Scenario
1. I have a _long_ glob of file names I want to **include** for formatting. 
2. I also have a _long_ glob of file names I want to **exclude** for formatting. 

Both pre-commit hook packages (e.g. `lint-staged`, `precise-commits`, and npm scripts need this filepath information, as well as the normal config. 

Is it possible?

**Comment by lydell** [comment]
Can you show an example?

**Comment by lydell** [comment]
Since you listed so many things in `.prettierignore`, can't you use `prettier **/*.{js,jsx}`?

**Comment by kevinSuttle** [comment]
I _could_, but what about `lint-staged`? It [still requires the inclusion list](https://github.com/okonet/lint-staged#reuse-npm-script).

**Comment by lydell** [comment]
So ... just write the inclusion list for `lint-staged`? The issue seems to be about avoiding repetition and that list isn't repeated?

**Comment by iammerrick** [comment]
The trouble with configuring it there is peoples prettier editor tooling doesn't respect lint-staged narrowing. So it still leaves one wanting some way of specifying includes only...

**Comment by kevinSuttle** [comment]
Yes, what @iammerrick said. ☝️

**Comment by lydell** [comment]
But the editor plugins support .prettierignore?

Note: Don't take it as I'm trying very hard to shoot down everything you say. I'm honestly just trying to understand the problem fully so we know what to do.

**Comment by iammerrick** [comment]
Yes. :-) The Editor plugins run prettier, which then supports .prettierignore but by having include "outside" of prettier with no way of declarative configuration means the only way to have the great tooling experience is to maintain a prettier ignore that is the direct inverse of the files you provide prettier. 

> Note: Don't take it as I'm trying very hard to shoot down everything you say. I'm honestly just trying to understand the problem fully so we know what to do.

Not taking it that way at all, I realize open source is free and you owe me nothing. Appreciate your time and energy. 👍

**Comment by lydell** [comment]
I think I might know too little about lint-staged to be able to assess this properly.

One last try: As far as I can tell, `prettier --write foo.js` where `foo.js` is ignored by .prettierignore is a no-op. Would it be possible to tell lint-staged to work on _all_ .js files (even the ones you don't care about) and rely on prettier not doing anything on the ignored files?

**Comment by iammerrick** [comment]
Yes. I think that is the current status.

**Comment by lydell** [comment]
Is there anything wrong with that?

**Comment by iammerrick** [comment]
I think the only issue is that you'd need to maintain a long list of excludes depending on your adoption level. Ideally, a prettier ignore could be a:

```!<includes pattern>```

But I haven't had success defining excludes as an inverted white list when using the `{multiple,folder,expansion}`. The nice thing with that is every time something new is added you don't need to go update your ignore file.. That is very much a different issue though. :-)

**Comment by lydell** [comment]
I see what you mean. That’s the approach I’ve taken with ESLint in several projects: `eslint .` and a .eslintignore that needs to be kept up-to-date. I’ve also tried to first ignore everything and then un-ignore the things I care about, but always bumped into issues as well. I wonder if there’s an ESLint issue we could take experience from.

**Comment by jpierson** [comment]
I'm seeing the same potential problem looking to adopt prettier in a project where we only manage a small set of sub-directories of the main project. A number of the developers are using VSCode but also other editors and wish to use the formatOnSave option to have prettier format using the locally installed version of prettier. The problem then is that in order to get VSCode pick up the locally installed version of prettier we have to install it at the root of the project which then makes prettier think that it should apply to every file recursively in the root. Then enters the task of configuring a number of editor extensions, pre-commit hooks, and other potential tools to ensure we don't apply formatting to other files outside of these sub-directories even if they need to be edited for some reason.

It seems life would be a lot easier if prettier just had a `files` glob list in it's configuration which when supplied would constrain prettier to formatting only files in this specified pattern. This way even if somebody edited a non included file in the editor, commited same file to git, or ran the CLI on the whole root directory it would be properly constrained by the config file to prevent formatting.

Conceptually I think what would work is a `.prettierinclude` file or at least the same concept baked into `.prettierrc` config.

**Comment by suchipi** [comment]
If you are slowly adopting prettier and have fewer files using it than not, then using [`requirePragma`](https://prettier.io/docs/en/options.html#require-pragma) might be the most ergonomic option.

**Comment by jpierson-at-riis** [comment]
@suchipi, in the case that I was mentioning we have thousands of files that this would apply to but it's a mixed code based repository with Java code and other sub projects which we are not ready to apply prettier on. By adding a subdirectory or part of the project at a time it may result in hundreds of files so `requirePrama` may be useful for some others who have few files that this may apply to.

Another point which I probably failed to mention is that the directories that we include are somewhat large due to the template structure used by the tooling which is out of our control. So in this case we would have something like show below in our `package.json` for making it easier to run prettier.

```json
{
    "scripts": {
        "prettier:enforce": "prettier \"abc-abc/web/src/wer/fwer/xsssf/src/js/something/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/xsssf/src/js/anotherthing/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/bbb/src/js/yetanother/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/xsssf/src/js/else/someother.css\"  --check",
        "prettier:fix": "prettier \"abc-abc/web/src/wer/fwer/xsssf/src/js/something/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/xsssf/src/js/anotherthing/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/bbb/src/js/yetanother/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/xsssf/src/js/else/someother.css\"  --write",
    }
}
```

**Comment by vincerubinetti** [comment]
> {
    "scripts": {
        "prettier:enforce": "prettier \"abc-abc/web/src/wer/fwer/xsssf/src/js/something/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/xsssf/src/js/anotherthing/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/bbb/src/js/yetanother/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/xsssf/src/js/else/someother.css\"  --check",
        "prettier:fix": "prettier \"abc-abc/web/src/wer/fwer/xsssf/src/js/something/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/xsssf/src/js/anotherthing/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/bbb/src/js/yetanother/**/*.{js,json,css,md}\" \"abc-abc/web/src/wer/fwer/xsssf/src/js/else/someother.css\"  --write",
    }
}

I was hoping to avoid this specific scenario (duplicating the long list of extensions, having to update both every time you make a change, leading to errors), and just be able to specify files in the prettier config file once.
